In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [10]:
#Importing ords_prods_merge.pkl data set from prepared data.
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')
ords_prods_merge = pd.read_pickle(file_path)

In [12]:
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   

   reordered  ... Unnamed: 0_y  product_name aisle_id  department_id  prices  \
0          0  ...          195          Soda       77              7     9.0   
1          1  ...          195          Soda       77              7     9.0   
2          1  ...          195          Soda       77              7     9.0   
3          1  ...          195          Soda       77              7     9.0   
4          1  ...          195          Soda       77              7     9.0   

   merge_flag    price_range_loc     busiest_day new_busiest_days  \
0        both  Mid-range product  Regularly busy   Regularly busy   
1        both  Mid-range product  Regularly busy     Slowest days   
2        both  Mid-range product  Regularly busy     Slowest days   
3        both  Mid-range product      Least busy     Slowest days   
4        both  Mid-range product      Least busy     Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  

[5 rows x 21 columns]

In [13]:
# Creating a subset df with the first one million entries
df = ords_prods_merge.head(1000000)

In [15]:
print(df.shape)

(1000000, 21)


In [16]:
#Grouping given dataframe by a given column ''product name''
df.groupby('product_name')

In [17]:
#Applying the agg() function to each group to obtain the mean values for the “order_number” column.
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [19]:
#Working with order number column, producing mean, min and max:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [20]:
#Creating the max order column.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\rbrand\AppData\Local\Temp\ipykernel_19384\1348309971.py:2: FutureWarning: The provided callable <function amax at 0x00000279F30DDBC0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [21]:
ords_prods_merge.head(15)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   

    order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                   8                   NaN         196                  1   
1                   7                  15.0         196                  1   
2                  12                  21.0         196                  1   
3                   7                  29.0         196                  1   
4                  15                  28.0         196                  1   
5                   7                  19.0         196                  1   
6                   9                  20.0         196                  1   
7                  14                  14.0         196                  2   
8                  16                   0.0         196                  4   
9                   8                  30.0         196                  1   
10                  9                   7.0         196                  2   
11                 16                   8.0         196                  6   
12                 11                   7.0         196                  1   
13                  9                   7.0         196                  1   
14                 10                  14.0         196                  1   

    reordered  ... product_name  aisle_id department_id  prices  merge_flag  \
0           0  ...         Soda        77             7     9.0        both   
1           1  ...         Soda        77             7     9.0        both   
2           1  ...         Soda        77             7     9.0        both   
3           1  ...         Soda        77             7     9.0        both   
4           1  ...         Soda        77             7     9.0        both   
5           1  ...         Soda        77             7     9.0        both   
6           1  ...         Soda        77             7     9.0        both   
7           1  ...         Soda        77             7     9.0        both   
8           1  ...         Soda        77             7     9.0        both   
9           1  ...         Soda        77             7     9.0        both   
10          0  ...         Soda        77             7     9.0        both   
11          1  ...         Soda        77             7     9.0        both   
12          1  ...         Soda        77             7     9.0        both   
13          1  ...         Soda        77             7     9.0        both   
14          1  ...         Soda        77             7     9.0        both   

      price_range_loc     busiest_day new_busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Regularly busy        Average orders   
1   Mid-range product  Regularly busy     Slowest days        Average orders   
2   Mid-range product  Regularly busy     Slowest days           Most orders   
3

Let’s run through each piece of the code in turn. First, a new column called “max_order” is created, which will be what stores the maximum order number for each user (step 3). Then, the ords_prods_merge dataframe is grouped by the “user_id” column (step 1). And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2). np = NumpPy library.

#Aggregation procedure checked with sintax: pd.options.display.max_rows = None. All works, goot tip for future reference.

In [27]:
#Creating flag to assign 'loyalty' label.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [28]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [29]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [33]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [34]:
#Checking that all the new flags were assigned correctly.
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [35]:
#7.Exporting the ords_prods_merge DataFrame as a Pickle file to reduce RAM Usage.
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [4]:
#Importing ords_prods_merge.pkl data set from prepared data.
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')
ords_prods_merge = pd.read_pickle(file_path)

-----------Exercise 4.8 ---------------------------------------------------------------------

In [5]:
#2.a Creating the subset for all the dataframe:
df = ords_prods_merge

In [6]:
print(df.shape)

(32404859, 23)


In [7]:
#2.b Grouping the entire dataframe by the "product_name" column:
df.groupby('product_name')

In [9]:
#2.c Applying the agg() function to obtain the mean values for the "order_number" column:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [8]:
#2.d Working with the "order_number" column, producing mean, min, and max for the entire dataframe:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [10]:
print(df.shape)
print(df.head())

(32404859, 23)
   Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  order_frequency_days  product_id  add_to_cart_order  \
0                  8                   NaN         196                  1   
1                  7                  15.0         196                  1   
2                 12                  21.0         196                  1   
3                  7                  29.0         196                  1   
4                 15                  28.0         196                  1   

   reordered  ... aisle_id  department_id prices  merge_flag  \
0          0  ...       7

#3. In this analysis, per 2nd request - I began by calculating the mean for all numeric columns in the entire dataset, providing insights into central tendencies across various features. Next, I explored the dataset by grouping it based on the "product_name" column, allowing me to investigate trends and patterns associated with specific products. Moving on, I calculated the mean order number for each department, offering a broad view of average order frequencies across different departments. Finally, I wnt a little deeper into the "order_number" column by computing additional statistics (mean, min, max) for each department taking as reference the first exersice from the task. This detailed analysis provides a clear understanding of order patterns within each department, contributing to a comprehensive exploration of the entire dataset and revealing valuable insights into product and department-wise dynamics.

In [11]:
#4a.Spliting the data into groups based on the “user_id” column.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\rbrand\AppData\Local\Temp\ipykernel_29400\3316690308.py:2: FutureWarning: The provided callable <function amax at 0x0000027E09021C60> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [12]:
#4b.Creating a new column, “max_order,” into which you’ll place the results of your aggregation.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [13]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [14]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [15]:
#4c.Checking that all the new flags were assigned correctly.
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(20)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10

In [16]:
print(df.shape)

(32404859, 23)


In [17]:
#5.a Grouping the data by loyalty_flag and calculating basic statistics for product prices for marketing team insight request.
loyalty_price_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

I'm using this code where the groupby() function is to group the data by the "loyalty_flag" column and then calculates basic statistics (mean, standard deviation, minimum, 25th percentile, median, 75th percentile, and maximum) for the "prices" column within each loyalty category using the describe() function learned on the past sessions. The results will provide insights to the marketing team into whether there's a difference in the spending habits of loyal customers compared to regular or new customers based on product prices.

In [18]:
#5.b Checking the basic statistics for product prices by loyalty category for insight.
print(loyalty_price_stats)

                       count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0  


#5c. In analyzing the spending habits of different customer types per marketing team request – loyal, new, and regular – some key observations emerge. 

- Firstly, there's a noticeable difference in the count of entries for each customer category, with loyal customers having the most, followed by regular customers, and then new customers. 

- Moving on to average prices, loyal customers spend approximately $10.39 on average, while new customers tend to spend a bit more at $13.29, and regular customers fall in between at around $12.50.

- Considering variation in spending, both loyal and new customers exhibit significant fluctuations in their purchase amounts, indicating diverse spending patterns. Regular customers also show some variation, but it's comparatively less than the other groups. Examining the price range, all customer types share the same minimum product price of $1.00. The prices at the middle percentiles (25th, 50th, and 75th) are quite similar across all customer categories. Surprisingly, the maximum product price is an unusually high $99,999.00, consistent across all customer types.

In summary, while the average prices are relatively close for all customer types, there's a considerable range in spending among loyal and new customers. The presence of the same maximum price for all customer types is noteworthy, though it raises some questions that may warrant further investigation. Additionally, the disparity in entry counts suggests that more data is available for loyal customers, which could influence our conclusions and should be taken into account in the analysis.

In [19]:
#6a.Calculating the mean prices for each user per team request
user_mean_prices = ords_prods_merge.groupby('user_id')['prices'].mean()

In [20]:
print(user_mean_prices.head())

user_id
1    6.367797
2    7.515897
3    8.197727
4    8.205556
5    9.189189
Name: prices, dtype: float64


In [21]:
#6b.Creating a spending flag based on the mean prices criteria per team request
ords_prods_merge['spending_flag'] = np.where(ords_prods_merge['user_id'].map(user_mean_prices) < 10, 'Low spender', 'High spender')

In [22]:
print(ords_prods_merge[['user_id', 'spending_flag']].head(10))

   user_id spending_flag
0        1   Low spender
1        1   Low spender
2        1   Low spender
3        1   Low spender
4        1   Low spender
5        1   Low spender
6        1   Low spender
7        1   Low spender
8        1   Low spender
9        1   Low spender


In [23]:
#6c.Checking the distribution of values in the spending_flag column
print(ords_prods_merge['spending_flag'].value_counts())

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64


Created the spending flag for each user based on the average price across all their orders to verify the results. It's evident that a significant portion of users are categorized as "Low spenders," with a count of 31,770,614, compared to "High spenders," which have a count of 634,245. This distribution suggests that the majority of users tend to make purchases with an average product price lower than 10, aligning with the criteria provided in the exercise #6 for being classified as a "Low spender." The marketing team can use this information to tailor their strategies and campaigns, potentially focusing on different approaches for these two distinct groups of users

In [24]:
#Exporting the ords_prods_merge DataFrame as a Pickle file to reduce RAM Usage.
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [3]:
#Importing ords_prods_merge.pkl data set from prepared data.
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')
ords_prods_merge = pd.read_pickle(file_path)

In [21]:
#7a. Calculating the median of order_frequency_days for each user per request. 
#Using 'order_frequency_days' since was renamed on 4.4 task.
median_order_frequency = ords_prods_merge.groupby('user_id')['order_frequency_days'].median()

In [22]:
#7b.Creating a new column 'order_frequency_flag' with default value ''
ords_prods_merge['order_frequency_flag'] = ''

In [24]:
#7c. Applying the criteria and label for customers per request using new isin method due a research.
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(median_order_frequency[median_order_frequency > 20].index), 'order_frequency_flag'] = 'Non-frequent customer'

In [26]:
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(median_order_frequency[(median_order_frequency > 10) & (median_order_frequency <= 20)].index), 'order_frequency_flag'] = 'Regular customer'

In [27]:
ords_prods_merge.loc[ords_prods_merge['user_id'].isin(median_order_frequency[median_order_frequency <= 10].index), 'order_frequency_flag'] = 'Frequent customer'

I used the "isin" method per my research to organize customers based on how often they order. Quick and smart tool that helped me check if a customer is in the group of those who order a lot, a medium amount, or not so often. This made things simpler and easier to understand in my code. "isin" made it possible to label customers without using complex steps, and it was like a friendly helper for handling a bunch of data.

In [28]:
# Step 4: Display the counts for each order frequency category
order_frequency_counts = ords_prods_merge['order_frequency_flag'].value_counts()
order_frequency_counts

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
                                5
Name: count, dtype: int64

The output indicates the count of customers in each category based on their order frequency. In your case: Where the breakdown provides insights into the distribution of customers based on their ordering behavior per instacart team request in the last #7 request.

In [29]:
#Exporting the ords_prods_merge DataFrame as a Pickle file.
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [1]:
#creating path
path = r'C:\Users\rbrand\12-28-2024 Instacart Basket Analysis'

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#Importing ords_prods_merge.pkl data set from prepared data.
file_path = os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl')
ords_prods_merge = pd.read_pickle(file_path)

In [4]:
#marking balues as missing since they doint make sens in terms of the other values in the column
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [5]:
#Checking if the replacement was succesfull
ords_prods_merge['prices'].max()

25.0

In [6]:
#Exporting the ords_prods_merge DataFrame as a Pickle file.
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))